keps/sig-network/4427-relaxed-dns-search-validation/README.md

```diff
@@ -0,0 +1,813 @@
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+Currently, Kubernetes applies the same restrictions to DNS search strings as it does to object names.
+These restrictions are derived from [RFC-1123](https://datatracker.ietf.org/doc/html/rfc1123). However,
+DNS name segments may contain more characters than are allowed in host names (see [RFC-2181](https://datatracker.ietf.org/doc/html/rfc2181#section-11)).
+In practice, DNS names sometimes contain characters, such as underscores, which would not be allowed by RFC-1123.
+
+This KEP proposes relaxing the checks on DNS search strings, specifically, without changing the restrictions on Kubernetes object
+names or other uses of DNS names within Kubernetes. Allowing these values in the `searches` field of `dnsConfig` allows pods to
+resolve short names properly in cases where the search string contains characters that are not currently allowed.
+
+## Motivation
+
+For workloads that resolve short DNS names where the full DNS name includes disallowed characters (like underscores),
+it’s not possible to configure search strings using dnsConfig. For example, if a pod needs to resolve abc123.abc_d.example.com
+using the short name of abc123, we would like to be able to add abc_d.example.com to the searches in the dnsConfig.
+
+Here’s an example configuration which would support this case:
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  namespace: default
+  name: dns-example
+spec:
+  containers:
+    - name: test
+      image: nginx
+  dnsPolicy: "None"
+  dnsConfig:
+    nameservers:
+      - 1.2.3.4
+    searches:
+      - abc_d.example.com
+```
+
+However, this returns an error:
+
+```
+The Pod "dns-example" is invalid: spec.dnsConfig.searches[0]: Invalid value: "abc_d.example.com": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
+```
+
+While a case could be made that these names are incorrect as host names based on RFC-1123, DNS servers generally allow such names
+and once they are in use DNS names can be complicated to change especially when they're used by legacy systems and not purely
+within Kubernetes. The DNS searches in the Kubernetes pods spec are not defining the host name, so even under the stricter interpretation
+of RFC-1123, relaxing these checks does not allow users of Kubernetes to introduce an invalid host name where one did not
+already exist outside of Kubernetes.
+
+### Goals
+
+- Support workloads that need to resolve DNS short names where the full DNS name includes characters disallowed by RFC-1123.
+
+## Proposal
+
+Introduce a RelaxedDNSSearchValidation feature gate which is disabled by default. When the feature gate is enabled,
+DNS search strings will be validated according to RFC-2181:
```

Comment:
See my comment here: https://github.com/kubernetes/enhancements/pull/4428#discussion_r1470391696 

I think the thing that justifies treating search strings differently is that they are not hostnames and the names formed from them are not required to be hostnames.

Reply:
I appreciate Dan's perspective here - if we JUST allow `'_'` then we have a much better sense of what's going to happen.  Is there any real use case for anything other than underscore?

The counter-argument is GIGO, which _seems_ to absolve us of the responsibility, but I'm not so sure.  Given that this is end-user controlled data which is examined by system components, we're still on the hook if something goes wrong.  IOW, it's a possible attack vector - do we really need to open it wider? 

The extension of this loosening to "all referred-to hostnames" seems out of scope for this KEP - I can't recall anyone evr asking for underscores in LB status or Ingress hosts, for example.  But if they did, I think it would be OK.  @robscott @shaneutt has this ever come up in Gateway?

Reply:
The path is fairly direct from dnsConfig -> resolv.conf -> resolver, where these characters are valid for DNS names (so are passed through there).

I'm not aware of any use-case beyond `_`, though, I'd be ok with limiting it to that..

Reply:
I think we can start with adding underscore for this specific validation alone.  Call it "DNS expanded" or something.  Then we can consider if others should also allow it later.  More surgical == more likely to fly.

Reply:
I'm not aware of any requests like this in Gateway API - currently the API is relatively underspecified when it comes to DNS, but our [relevant validation](https://github.com/kubernetes-sigs/gateway-api/blob/1edf7e93ae4bb8c5ea5cc2193586cfe1e5d15e2f/apis/v1/shared_types.go#L540) follows RFC 1035 and does not allow underscores for what it's worth. With that said, it seems reasonable in this limited context to allow underscores, it's just a bit of a slippery slope we're starting down.

Reply:
I wouldn't without real use-cases.  TLS certs, for example, are not supposed to include domain names with underscores.